### PR TITLE
Adds random within range '~' operator

### PR DESCRIPTION
**Pull request by opohorel**
I've got this patch from Matej Mužila. It implements '~' operator that is discussed in #85 .

I've tested this patch and it seems like it works as expected, however I suggest that other people test it as well and provide me with some feedback.

This is a diagram showing how entry.c now works that Matej provided with this patch.
fsm-1

### CODE REVIEW

**Comment by t8m** [comment]
Nice, could you please add a documentation?

**Comment by opohorel** [comment]
> Nice, could you please add a documentation?

I've added a description into man/crontab.5. Please take a look if anything is missing.

**Comment by sgerwk** [comment]
Now it works.

I tried an actual random range, this way:

```
1~10 * * * * date >> /tmp/log
```

To save time, instead of letting crond run for hours, I run it in simulated
time, under [timeserver](https://github.com/sgerwk/timeserver):

```
timeserver
timeexec ./src/crond -n -x ext,sch,proc,pars,load,misc,bit -m"tail -1"
timerun $((20000))
kill -HUP (pid of crond)
timerun $((20000))
timerun $((20000))
```

I first started the simulated time server in a terminal, then crond in another
terminal, and then I run the simulation for 20000 seconds, then send a SIGHUP
to crond, run for other 20000 seconds and yet other 20000 seconds.

At the end,  /tmp/log contains:

```
Thu Jan  1 01:05:24 CET 1970
Thu Jan  1 02:05:23 CET 1970
Thu Jan  1 03:05:16 CET 1970
Thu Jan  1 04:05:22 CET 1970
Thu Jan  1 05:05:20 CET 1970
Thu Jan  1 06:05:23 CET 1970
Thu Jan  1 07:04:19 CET 1970
Thu Jan  1 08:04:18 CET 1970
Thu Jan  1 09:04:23 CET 1970
Thu Jan  1 10:04:17 CET 1970
Thu Jan  1 11:04:22 CET 1970
Thu Jan  1 12:04:24 CET 1970
Thu Jan  1 13:04:18 CET 1970
Thu Jan  1 14:04:19 CET 1970
Thu Jan  1 15:04:21 CET 1970
Thu Jan  1 16:04:22 CET 1970
Thu Jan  1 17:04:27 CET 1970
```

It works, but not the way I would consider "running at random time". I would
expect the number of the minute to change every time. Instead, it is chosen
once at the beginning, and only changed when crond receives a SIGHUP. In other
words, the random time is set or changed only when crond reads the crontable.

I am not sure this is what one wants from random time. Actually, it's not clear
to me when such a feature could be useful. Could you provide an example?

**Comment by opohorel** [comment]
> Now it works.
> 

Great! Thank you for testing this!

> It works, but not the way I would consider "running at random time". I would expect the number of the minute to change every time. Instead, it is chosen once at the beginning, and only changed when crond receives a SIGHUP. In other words, the random time is set or changed only when crond reads the crontable.
> 
> I am not sure this is what one wants from random time. Actually, it's not clear to me when such a feature could be useful. Could you provide an example?

This has been described in #85. To be exact, it was mentioned in [this mail](https://marc.info/?l=openbsd-tech&m=158683644314416&w=2). The random time changes when parsing the crontab and stays until it is parsed again, which typically happens when crond is reloaded.

I'm trying to implement this feature on behalf of a customer. I don't use cronie that much, that's why I can't give you the exact example when it is useful.

**Comment by t8m** [comment]
I think it is useful even as implemented. The main reason for the randomization is to avoid mass mailing or power consumption on the same time synchronized by the cron table entries when there are multiple virtual or physical machines with the same cron tables in a datacenter. It does not really matter whether the time is randomized only once on the machine or every time the cron job is supposed to be executed.

Implementing it the other way would be also much more complicated and in some cases inconvenient because it would mean the time between the individual executions of the job would vary.

File: src/entry.c [hunk 1]
```diff
@@ -486,144 +502,193 @@ get_list(bitstr_t * bits, int low, int high, const char *names[],
...
+					num1 = num2 = -1;
 
-	/* num1 (through i) will be validated by set_element() below, but num2
-	 * and num3 are merely used as loop condition and increment, and must
-	 * be validated separately.
-	 */
-	if (num2 < low || num2 > high || num3 > high)
+				/* select random number in range <num1, num2>
+				 * */
+				num1 = num2 = random() % (num2 - num1 + 1) + num1;
```
Comment (t8m) [inline src/entry.c:633]: This is biased rng. Would  you please consider https://codereview.stackexchange.com/questions/159604/uniform-random-numbers-in-an-integer-interval-in-c

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by sgerwk** [comment]
> I think it is useful even as implemented. The main reason for the randomization is to avoid mass mailing or power consumption on the same time synchronized by the cron table entries when there are multiple virtual or physical machines with the same cron tables in a datacenter. It does not really matter whether the time is randomized only once on the machine or every time the cron job is supposed to be executed.
> 
> Implementing it the other way would be also much more complicated and in some cases inconvenient because it would mean the time between the individual executions of the job would vary.

That makes sense. I see it is also documented this way: "Random number is picked when crontab file is parsed." Perhaps this could be further clarified by "which happens when crond starts and when it receives a SIGHUP"?

File: src/entry.c (hunk 1 above)
Comment (opohorel) [reply src/entry.c:633]: I'm not sure that it is necessary to make rng unbiased in this use case. Do you insist on implementing this feature?

Comment (t8m) [reply src/entry.c:633]: OK, it's true that the bias won't be big enough to matter given the range is always fairly short in comparison to the random() output range.

File: src/entry.c [hunk 2]
```diff
@@ -62,9 +62,22 @@ static const char *ecodes[] = {
...
+	R_AST,
+	R_STEP,
+	R_TERMS,
+	R_NUM1,
+	R_RANGE,
+	R_RANGE_NUM2,
+	R_RANDOM,
+	R_RANDOM_NUM2,
+	R_FINISH,
+}range_state_t;
```
Comment (t8m) [inline src/entry.c:76]: Please add space after }

File: man/crontab.5 [hunk 3]
```diff
@@ -205,6 +205,15 @@ hyphen.  The specified range is inclusive.  For example, 8-11 for
 an 'hours' entry specifies execution at hours 8, 9, 10, and 11. The first
 number must be less than or equal to the second one.
 .PP
+Randomization of the execution time within a range can be used.
+A random number within a range specified as two numbers separated with
+a tilde is picked.  The specified range is inclusive.
+For example, 6~15 for a 'minutes' entry picks a random minute
+within 6 to 15 range.  The random number is picked when crontab file is parsed.
+The first number must be less than or equal to the second one. You might omit
+the numbers specifying the range.  For example, ~ for a 'minutes' entry picks
```
Comment (t8m) [inline man/crontab.5:214]: Please make it more explicit: `You might omit one or both numbers specifying the range.`

File: src/entry.c [hunk 4]
```diff
@@ -486,144 +502,193 @@ get_list(bitstr_t * bits, int low, int high, const char *names[],
 	return (ch);
 }
 
+inline static int is_separator(int ch) {
+	switch(ch) {
```
Comment (t8m) [inline src/entry.c:506]: Add space after `switch`

File: src/entry.c [hunk 5]
```diff
@@ -486,144 +502,193 @@ get_list(bitstr_t * bits, int low, int high, const char *names[],
...
+	int ch, i, num1, num2, num3;
 
-	int i, num1, num2, num3;
+	/* default value for step
+	 */
+	num3 = 1;
+	range_state_t state = R_START;
+
+	while (state != R_FINISH && ((ch = get_char(file)) != EOF)) {
+		switch(state) {
```
Comment (t8m) [inline src/entry.c:534]: Add space after switch.

**Review by t8m — CHANGES_REQUESTED** [review]
Just a few more nits and I'll merge. Are you OK if I squash the commits when merging?

**Comment by opohorel** [comment]
Of course you can squash these commits. I just wanted to follow good git practices :)

**Pull request merged**